File: frontend/src/utils/riskCalculator.js
// Risk calculation utilities
// These functions implement the TILT vulnerability scoring logic
// Note: Some calculations are now done on backend, but we keep these for client-side preview

/**
 * Calculate vulnerability percentage based on assessment scores
 * @param {Object} answers - Object mapping question IDs to maturity scores (1-5)
 * @param {Array} questions - Array of question objects with weights
 * @param {Object} osintResults - OSINT check results { ransomware: boolean, ssl: string, dmarc: boolean }
 * @returns {number} Vulnerability percentage (0-100)
 */
export function calculateVulnerability(answers, questions, osintResults = null) {
  // No answers means we have no data, assume worst case
  if (!answers || Object.keys(answers).length === 0) {
    return 100;
  }
  
  let totalWeight = 0;
  let weightedScore = 0;
  
  // Calculate weighted average of maturity scores
  // Maturity scale: 1 = worst, 5 = best
  // We invert it so 1 becomes 1.0 (100% vulnerable) and 5 becomes 0.2 (20% vulnerable)
  questions.forEach(q => {
    const answer = answers[q.id];
    if (answer !== undefined && answer !== null) {
      const normalizedScore = (6 - answer) / 5; // Invert: 1->1.0, 5->0.2
      
      // Weight by importance (Critical questions matter 3x more than Low)
      const weight = q.weight === 'Critical' ? 3 : q.weight === 'High' ? 2 : 1;
      
      weightedScore += normalizedScore * weight;
      totalWeight += weight;
    }
  });
  
  if (totalWeight === 0) {
    return 100;
  }
  
  // Convert to percentage (0-100)
  let vuln = (weightedScore / totalWeight) * 100;
  
  // Apply OSINT adjustments to vulnerability
  // SSL and DMARC issues indicate poor security posture and increase vulnerability
  if (osintResults) {
    if (osintResults.ssl === 'F' || osintResults.ssl === 'Expired') {
      // Expired SSL is a serious security issue
      vuln = Math.max(vuln, 80);
    }
    if (osintResults.dmarc === false) {
      // Missing DMARC indicates email security weakness
      vuln = Math.max(vuln, 60);
    }
  }
  
  return Math.round(vuln);
}

/**
 * Calculate vulnerability for cloud scorecard
 * @param {Object} scores - Object mapping criterion IDs to maturity scores (0-5)
 * @param {Array} criteria - Array of cloud criteria objects
 * @param {Object} osintResults - OSINT check results { ransomware: boolean, ssl: string, dmarc: boolean }
 * @returns {Object} { score, maxScore, vulnerability, status }
 */
export function calculateCloudVulnerability(scores, criteria, osintResults = null) {
  // Calculate weighted score: maturity score (0-5) * points for that question
  const totalScore = criteria.reduce((sum, item) => {
    const maturityScore = scores[item.id] || 0; // Default to 0 if not answered
    const weightedPoints = (maturityScore / 5) * (item.points || 0);
    return sum + weightedPoints;
  }, 0);
  
  const maxScore = 55; // Total points available (all questions at level 5)
  
  // Check for critical failures: Critical questions with score 0
  const missedCritical = criteria.some(c => 
    c.criticality === 'Critical' && (scores[c.id] === 0 || scores[c.id] === undefined || scores[c.id] === null)
  );
  
  // Calculate vulnerability (inverse of score)
  // Higher score = lower vulnerability
  let vuln = Math.round((1 - (totalScore / maxScore)) * 100);
  
  // Apply OSINT adjustments to vulnerability
  // SSL and DMARC issues indicate poor security posture and increase vulnerability
  if (osintResults) {
    if (osintResults.ssl === 'F' || osintResults.ssl === 'Expired') {
      // Expired SSL is a serious security issue
      vuln = Math.max(vuln, 80);
    }
    if (osintResults.dmarc === false) {
      // Missing DMARC indicates email security weakness
      vuln = Math.max(vuln, 60);
    }
  }
  
  // Determine status based on new thresholds
  // 45-55: Elite/Low Risk (Pre-approved)
  // 30-44: Standard/Medium Risk (Standard Approval)
  // 15-29: Deficient/High Risk (Restricted)
  // 0-14: Critical Risk (REJECT)
  let status = 'Elite (Pre-approved)';
  let statusColor = 'green';
  
  if (missedCritical || totalScore < 15) {
    status = 'REJECT (Critical Risk)';
    statusColor = 'red';
  } else if (totalScore < 30) {
    status = 'Restricted (Deficient)';
    statusColor = 'amber';
  } else if (totalScore < 45) {
    status = 'Standard Approval';
    statusColor = 'yellow';
  } else {
    status = 'Elite (Pre-approved)';
    statusColor = 'green';
  }
  
  return {
    score: Math.round(totalScore),
    maxScore,
    vulnerability: vuln,
    status,
    statusColor,
    missedCritical,
  };
}

/**
 * Calculate inherent risk (Impact x Likelihood)
 * Simple multiplication for the risk matrix
 * @param {number} impact - Impact score (1-5)
 * @param {number} likelihood - Likelihood score (1-5)
 * @returns {number} Risk score
 */
export function calculateInherentRisk(impact, likelihood) {
  return impact * likelihood;
}

/**
 * Get risk zone color based on impact and likelihood
 * Used for the risk matrix visualisation
 * @param {number} impact - Impact score (1-5)
 * @param {number} likelihood - Likelihood score (1-5)
 * @returns {string} Color hex code
 */
export function getRiskZoneColor(impact, likelihood) {
  const risk = calculateInherentRisk(impact, likelihood);
  
  // Risk zones: 20+ = red, 12-19 = orange, 6-11 = yellow, <6 = green
  if (risk >= 20) return '#ef4444'; // Red
  if (risk >= 12) return '#f97316'; // Orange
  if (risk >= 6) return '#eab308';   // Yellow
  return '#22c55e'; // Green
}

File: backend/risk_calculator.py
# Risk calculation logic - Python re-implementation
# Mirrors the JavaScript logic from frontend/src/utils/riskCalculator.js
# X=Impact, Y=Likelihood, Color=Vulnerability per TILT framework

from typing import Dict, List, Any, Optional

def calc_vulnerability(answers: Dict[int, int], questions: List[Dict], osint_results: Optional[Dict] = None) -> int:
    """
    Calculate vulnerability percentage based on assessment scores.
    Returns 0-100 where 100 = fully vulnerable.
    """
    # Sanity check
    if not answers or len(answers) == 0:
        return 100
    
    total_weight = 0
    weighted_score = 0
    
    # Calculate weighted average of maturity scores
    # Maturity scale: 1 = worst, 5 = best
    # Invert: 1->1.0 (100% vulnerable), 5->0.2 (20% vulnerable)
    for q in questions:
        qid = q['id']
        if qid in answers and answers[qid] is not None:
            answer_val = answers[qid]
            normalized = (6 - answer_val) / 5  # Invert: 1->1.0, 5->0.2
            
            # Weight by importance (Critical = 3x, High = 2x, Low = 1x)
            weight = 3 if q.get('weight') == 'Critical' else (2 if q.get('weight') == 'High' else 1)
            
            weighted_score += normalized * weight
            total_weight += weight
    
    if total_weight == 0:
        return 100
    
    # Convert to percentage
    vuln = (weighted_score / total_weight) * 100
    
    # Apply OSINT adjustments to vulnerability
    # SSL and DMARC issues indicate poor security posture and increase vulnerability
    if osint_results:
        if osint_results.get('ssl') in ('F', 'Expired'):
            # Expired SSL is a serious security issue
            vuln = max(vuln, 80)
        if osint_results.get('dmarc') is False:
            # Missing DMARC indicates email security weakness
            vuln = max(vuln, 60)
    
    return round(vuln)

def calc_cloud_vulnerability(scores: Dict[int, int], criteria: List[Dict], osint_results: Optional[Dict] = None) -> Dict[str, Any]:
    """
    Calculate vulnerability for cloud scorecard.
    Scores are maturity levels (0-5) per question.
    Returns dict with score, maxScore, vulnerability, status, statusColor, missedCritical.
    """
    # Calculate weighted score: maturity score (0-5) * points for that question
    total_score = 0
    for c in criteria:
        maturity_score = scores.get(c['id'], 0)  # Default to 0 if not answered
        points = c.get('points', 0)
        weighted_points = (maturity_score / 5) * points
        total_score += weighted_points
    
    max_score = 55  # Total points available (all questions at level 5)
    
    # Check for critical failures: Critical questions with score 0
    missed_critical = any(
        c.get('criticality') == 'Critical' and 
        (scores.get(c['id']) == 0 or scores.get(c['id']) is None)
        for c in criteria
    )
    
    # Calculate vulnerability (inverse of score)
    vuln = round((1 - (total_score / max_score)) * 100) if max_score > 0 else 100
    
    # Apply OSINT adjustments to vulnerability
    # SSL and DMARC issues indicate poor security posture and increase vulnerability
    if osint_results:
        if osint_results.get('ssl') in ('F', 'Expired'):
            # Expired SSL is a serious security issue
            vuln = max(vuln, 80)
        if osint_results.get('dmarc') is False:
            # Missing DMARC indicates email security weakness
            vuln = max(vuln, 60)
    
    # Determine status based on new thresholds
    # 45-55: Elite/Low Risk (Pre-approved)
    # 30-44: Standard/Medium Risk (Standard Approval)
    # 15-29: Deficient/High Risk (Restricted)
    # 0-14: Critical Risk (REJECT)
    if missed_critical or total_score < 15:
        status = 'REJECT (Critical Risk)'
        status_color = 'red'
    elif total_score < 30:
        status = 'Restricted (Deficient)'
        status_color = 'amber'
    elif total_score < 45:
        status = 'Standard Approval'
        status_color = 'yellow'
    else:
        status = 'Elite (Pre-approved)'
        status_color = 'green'
    
    return {
        'score': round(total_score),
        'maxScore': max_score,
        'vulnerability': vuln,
        'status': status,
        'statusColor': status_color,
        'missedCritical': missed_critical,
    }

def calc_impact_likelihood(answers: Dict[int, int], questions: List[Dict], osint_results: Optional[Dict] = None) -> tuple:
    """
    Calculate impact and likelihood scores (1-5) from assessment answers.
    Returns (impact, likelihood) tuple.
    """
    # Default values
    impact = 3
    likelihood = 3
    
    if answers and len(answers) > 0:
        # Convert average maturity score to impact/likelihood
        # Higher maturity = better security = lower risk
        answer_vals = [v for v in answers.values() if v is not None]
        if answer_vals:
            avg_score = sum(answer_vals) / len(answer_vals)
            # Invert: high score = low impact/likelihood
            impact = max(1, min(5, 6 - avg_score / 1.2))
            likelihood = max(1, min(5, 6 - avg_score / 1.2))
    
    # OSINT adjustment: Only ransomware affects likelihood
    # DMARC and SSL issues are informational warnings only - they do NOT affect likelihood
    if osint_results and osint_results.get('ransomware') is True:
        likelihood = 5
    
    # Note: DMARC warnings are informational only and do not affect likelihood calculation
    # SSL issues may affect vulnerability percentage but not likelihood
    
    return (impact, likelihood)

